### PR TITLE
[Improvement] A tool for converting the format of proposal file.

### PR DESCRIPTION
**Pull request by tony2016uestc**

### CODE REVIEW

File: tools/data/activitynet/convert_proposal_format.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        if best_iou[i] == 0 and best_overlap[i] == 0:
+            action_classifier_input.write(
+                f'0 0 0 {proposals[i][0]} {proposals[i][1]}\n')
+        else:
+            action_classifier_input.write(
+                f'{int(label)} {best_iou[i]} {best_overlap[i]} '
+                f'{proposals[i][0]} {proposals[i][1]}\n')
+
+
+data_path = '../../../data/ActivityNet'
```
Comment (SuX97) [inline tools/data/activitynet/convert_proposal_format.py:98]: use arg parser to specify the path of output `action_classifier_input_file ,  input `proposal_generator_output_file `, category file `activity_index_file ` and annotation. Better not hard code it.

File: tools/data/activitynet/convert_proposal_format.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        else:
+            label = label_iou
+        if best_iou[i] == 0 and best_overlap[i] == 0:
+            action_classifier_input.write(
+                f'0 0 0 {proposals[i][0]} {proposals[i][1]}\n')
+        else:
+            action_classifier_input.write(
+                f'{int(label)} {best_iou[i]} {best_overlap[i]} '
+                f'{proposals[i][0]} {proposals[i][1]}\n')
+
```
Comment (SuX97) [inline tools/data/activitynet/convert_proposal_format.py:96]: Wrap it into a main function.

File: tools/data/activitynet/convert_proposal_format.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+data_path = '../../../data/ActivityNet'
+result_path = '../../..'
+proposal_generator_output_file = f'{result_path}/results.json'
+action_classifier_input_file = f'{result_path}/anet_val_classifier_input.txt'
+activity_index_file = f'{data_path}/anet_activity_indexes_val.txt'
+ann_file = f'{data_path}/anet_anno_val.json'
+
+action_classifier_input = open(action_classifier_input_file, 'w')
```
Comment (SuX97) [inline tools/data/activitynet/convert_proposal_format.py:105]: Still, do not call it `xxx_input`, it is the input of another module, but the output of this file. Called it `formatted_proposal_file`

File: tools/data/activitynet/convert_proposal_format.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                f'0 0 0 {proposals[i][0]} {proposals[i][1]}\n')
+        else:
+            action_classifier_input.write(
+                f'{int(label)} {best_iou[i]} {best_overlap[i]} '
+                f'{proposals[i][0]} {proposals[i][1]}\n')
+
+
+data_path = '../../../data/ActivityNet'
+result_path = '../../..'
+proposal_generator_output_file = f'{result_path}/results.json'
```
Comment (SuX97) [inline tools/data/activitynet/convert_proposal_format.py:100]: Just `proposal_file`, being consisit with `formatted_proposal_file`

File: tools/data/activitynet/convert_proposal_format.py [hunk 5]
```diff
@@ -87,46 +91,87 @@ def dump_classifier_input(video_idx, video_id, num_frames, fps, gts, proposals,
...
+        'name of formatted proposal file, which is the input of action classifier (SSN)'
+    )
+    args = parser.parse_args()
+
+    return args
+
+
+if __name__ == '__main__':
+    args = parse_args()
+    proposal_file = osp.join(args.result_path, args.proposal_file)
```
Comment (SuX97) [inline tools/data/activitynet/convert_proposal_format.py:143]: Do not need to split it into dir and filename. Just use `proposal_file` as the path.

File: tools/data/activitynet/convert_proposal_format.py [hunk 6]
```diff
@@ -87,46 +91,87 @@ def dump_classifier_input(video_idx, video_id, num_frames, fps, gts, proposals,
...
+    formatted_proposal_file = open(
+        osp.join(args.result_path, args.formatted_proposal_file), 'w')
+    ann_file = osp.join(args.data_path, args.ann_file)
+    activity_index_file = osp.join(args.data_path, args.activity_index_file)
```
Comment (SuX97) [inline tools/data/activitynet/convert_proposal_format.py:147]: Same, just use the path of the file.

**Comment by SuX97** [comment]
LGTM. @dreamerlin

**Review by SuX97 — APPROVED** [review]

File: configs/localization/bmn/README.md [hunk 7]
```diff
@@ -70,5 +70,10 @@ You can also test the action detection performance of the model, with [anet_cuhk
 python tools/analysis/report_map.py --proposal path/to/proposal_file
 ```
 
+Notes:
+1. (Optional) You can use the following command to generate a formatted proposal file, which will be fed into the action classifier (e.g: SSN) to get the classification result of proposals.
```
Comment (innerlee) [inline configs/localization/bmn/README.md:74]: What does `formatted proposal file` mean?

Comment (innerlee) [reply configs/localization/bmn/README.md:74]: Any other `action classifier` (e.g. I3D) could be used?

File: mmaction/core/evaluation/accuracy.py [hunk 8]
```diff
@@ -197,14 +197,18 @@ def binary_precision_recall_curve(y_score, y_true):
...
     Args:
         candidate_segments (np.ndarray): 1-dim/2-dim array in format
             ``[init, end]/[m x 2:=[init, end]]``.
         target_segments (np.ndarray): 2-dim array in format
             ``[n x 2:=[init, end]]``.
+        overlap_self (bool): Whether to calculate overlap_self
+            (union / candidate_length) or not. Default: False.
 
     Returns:
         t_iou (np.ndarray): 1-dim array [n] /
```
Comment (innerlee) [inline mmaction/core/evaluation/accuracy.py:214]: It may return result of `overlap_self`

File: mmaction/core/evaluation/accuracy.py [hunk 9]
```diff
@@ -197,14 +197,18 @@ def binary_precision_recall_curve(y_score, y_true):
...
+                          target_segments,
+                          overlap_self=False):
     """Compute intersection over union between segments.
 
     Args:
         candidate_segments (np.ndarray): 1-dim/2-dim array in format
             ``[init, end]/[m x 2:=[init, end]]``.
         target_segments (np.ndarray): 2-dim array in format
             ``[n x 2:=[init, end]]``.
+        overlap_self (bool): Whether to calculate overlap_self
```
Comment (innerlee) [inline mmaction/core/evaluation/accuracy.py:210]: `overlap_self` -> `return_overlap_self` or `calculate_overlap_self`

File: tools/data/activitynet/convert_proposal_format.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
+"""This file converts the output proposal file of proposal generator (BMN) into
+the input proposal file of action classifier (SSN)."""
```
Comment (innerlee) [inline tools/data/activitynet/convert_proposal_format.py:2]: Is this script restricted to BMN and SSN?

Comment (tony2016uestc) [reply tools/data/activitynet/convert_proposal_format.py:2]: No, they are just examples.

File: configs/localization/bmn/README.md (hunk 7 above)
Comment (tony2016uestc) [reply configs/localization/bmn/README.md:74]: Only SSN and PGCN use this `formatted proposal file` as input, and generate the classification result of proposals.

Comment (innerlee) [reply configs/localization/bmn/README.md:74]: Okay, you can put this info into the notes. When one see `action classifier`, he/she may think about TSN etc.

Comment (tony2016uestc) [reply configs/localization/bmn/README.md:74]: OK.

File: tools/data/activitynet/convert_proposal_format.py (hunk 10 above)
Comment (innerlee) [reply tools/data/activitynet/convert_proposal_format.py:2]: If possible, provide a full list of supported input/output methods.

Comment (tony2016uestc) [reply tools/data/activitynet/convert_proposal_format.py:2]: OK.

**Review by innerlee — APPROVED** [review]

**Pull request merged**